README.md
# Webdrivers

[![Gem Version](https://badge.fury.io/rb/webdrivers.svg)](https://badge.fury.io/rb/webdrivers)
![Tests](https://github.com/titusfortner/webdrivers/workflows/Tests/badge.svg)

Run Selenium tests more easily with automatic installation and updates for all supported webdrivers.

## Update: Future of this Project

With Google's new [Chrome for Testing](https://developer.chrome.com/blog/chrome-for-testing/) project, 
and Selenium's new [Selenium Manager](https://www.selenium.dev/documentation/selenium_manager/) feature, 
what is required of this gem has changed.

If you can update to the latest version of Selenium (4.11+), please do so and stop requiring this gem.
Provide feedback or raise issues to [Selenium Project](https://github.com/SeleniumHQ/selenium/issues/new/choose)

If you cannot upgrade to Selenium 4.11, Webdrivers 5.3.0 will continue to support Ruby 2.6+ and Selenium 4.0 - 4.10

If you are using an older version of webdrivers gem, and cannot upgrade, you can set
the required version of chromedriver to v114 (`Webdrivers::Chromedriver.required_version = '114.0.5735.90'`) and 
[Disable the build check](https://www.selenium.dev/documentation/webdriver/browsers/chrome/#disabling-build-check). 
This is not guaranteed to continue working and will not receive bug fixes.

**Webdrivers 6.0**
To provide support for Selenium 3 and Ruby < 2.6 a 6.0 version is planned. It requires:
* Creating a `selenium-manager.gem` based off of https://github.com/SeleniumHQ/selenium/pull/12429
* Re-implementing this gem to wrap `selenium-manager.gem`
* Ensuring compatible with older versions of Selenium & Ruby

If anyone would like to help get Webdrivers 6 working, please let us know.

## Description

`webdrivers` downloads drivers and directs Selenium to use them. Currently supports:

* [chromedriver](http://chromedriver.chromium.org/)
* [geckodriver](https://github.com/mozilla/geckodriver)
* [IEDriverServer](https://github.com/SeleniumHQ/selenium/wiki/InternetExplorerDriver)
* [msedgedriver](https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver/)

Works on macOS, Linux, Windows, and Windows Subsystem for Linux (WSL) v1 and v2. And do see the browser and OS specific
notes at the bottom.

## Usage

In your Gemfile:

```ruby
gem 'webdrivers', '~> 5.0', require: false
```

In your project:

```ruby
require 'webdrivers'
```

The drivers will now be automatically downloaded or updated when you launch a browser
through Selenium.

### Specific Drivers

If you want webdrivers to only manage specific drivers you can specify one or more as follows:

```ruby
require 'webdrivers/chromedriver'
require 'webdrivers/geckodriver'
require 'webdrivers/iedriver'
require 'webdrivers/edgedriver'
```

### Download Location

The default download location is `~/.webdrivers` directory, and this is configurable:

 ```ruby
 Webdrivers.install_dir = '/webdrivers/install/dir'
```

Alternatively, you can define the path via the `WD_INSTALL_DIR` environment
variable.

### Version Pinning

If you would like to use a specific (older or beta) version, you can specify it for each driver. Otherwise,
the latest (stable) driver will be downloaded and passed to Selenium.

```ruby
# Chrome
Webdrivers::Chromedriver.required_version = '2.46'

# Firefox
Webdrivers::Geckodriver.required_version  = '0.23.0'

# Internet Explorer
Webdrivers::IEdriver.required_version     = '3.14.0'

# Edge (Chromium)
Webdrivers::Edgedriver.required_version   = '76.0.183.0'
```

You can explicitly trigger the update in your code, but this will happen
automatically when the driver is initialized:

```ruby
Webdrivers::Chromedriver.update
```

### Caching Drivers

You can set Webdrivers to only look for updates if the previous check
was longer ago than a specified number of seconds.

```ruby
Webdrivers.cache_time = 86_400 # Default: 86,400 Seconds (24 hours)
```

Alternatively, you can define this value via the `WD_CACHE_TIME` environment
variable. **Only set one to avoid confusion**.

##### Special exception for chromedriver and msedgedriver

Cache time will be respected as long as a driver binary exists and the major.minor.build versions of
the browser and the driver match. For example, if you update Chrome or Edge to v76.0.123 and its driver is
still at v76.0.100, `webdrivers` will ignore the cache time and update the driver to make sure you're
using a compatible build version.

### Proxy

If there is a proxy between you and the Internet then you will need to configure
the gem to use the proxy.  You can do this by calling the `configure` method.

````ruby
Webdrivers.configure do |config|
  config.proxy_addr = 'myproxy_address.com'
  config.proxy_port = '8080'
  config.proxy_user = 'username'
  config.proxy_pass = 'password'
end
````

### `SSL_connect` errors

If you are getting an error like this (especially common on Windows):

`SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed`

Add the following to your Gemfile:

```ruby
gem "net_http_ssl_fix"
```

Add the following to your code:

````ruby
require 'net_http_ssl_fix'
````

Other solutions are documented on the RubyGems [website](https://guides.rubygems.org/ssl-certificate-update/).

### Rake tasks

Each driver has its own set of `rake` tasks (with `Railtie` support) that
you can call once before executing the tests. These are especially
useful if you're running tests in parallel and want to avoid performing
an update check per thread.

If you are using Rails default configuration the `webdrivers` gem will only be loaded in the test group
so you will need to specify the test environment when using the tasks:

```ruby
RAILS_ENV=test rails webdrivers:chromedriver:update
```

If you are not using Rails, you'll need to load them into your Rakefile like this:

```ruby
require 'webdrivers'
load 'webdrivers/Rakefile'
```

The full list of available tasks is:

```bash
$ bundle exec rake -T
rake webdrivers:chromedriver:remove           # Force remove chromedriver
rake webdrivers:chromedriver:update[version]  # Remove and download updated chromedriver if necessary
rake webdrivers:chromedriver:version          # Print current chromedriver version
rake webdrivers:edgedriver:remove             # Force remove msedgedriver
rake webdrivers:edgedriver:update[version]    # Remove and download updated msedgedriver if necessary
rake webdrivers:edgedriver:version            # Print current msedgedriver version
rake webdrivers:geckodriver:remove            # Force remove geckodriver
rake webdrivers:geckodriver:update[version]   # Remove and download updated geckodriver if necessary
rake webdrivers:geckodriver:version           # Print current geckodriver version
rake webdrivers:iedriver:remove               # Force remove IEDriverServer
rake webdrivers:iedriver:update[version]      # Remove and download updated IEDriverServer if necessary
rake webdrivers:iedriver:version              # Print current IEDriverServer version
```

These tasks respect the `WD_INSTALL_DIR`, `WD_CACHE_TIME`, `WD_CHROME_PATH`,
and `WD_EDGE_CHROME_PATH` environment variables, which can also be passed
through the `rake` command:

```bash
$ bundle exec rake webdrivers:chromedriver:update[2.46] webdrivers:geckodriver:update[0.24.0] WD_CACHE_TIME=86_400 WD_INSTALL_DIR='my_dir'
2019-05-20 19:03:01 INFO Webdrivers Updated to chromedriver 2.46.628388
2019-05-20 19:03:04 INFO Webdrivers Updated to geckodriver 0.24.0
```

Please note that these tasks do not use any of the configurations from your
project (code) and only respect the `ENV` variables and the version (optional)
passed to the `rake` tasks.

### Logging

The logging level can be configured for debugging purpose:

```ruby
Webdrivers.logger.level = :DEBUG
```

### Browser & OS Specific Notes

#### Chrome/Chromium

The version of `chromedriver` will depend on the version of Chrome you are using it with:

 * For versions >= 70, the downloaded version of `chromedriver` will match the installed version of Google Chrome.
 More information [here](http://chromedriver.chromium.org/downloads/version-selection).
 * For versions <=  69, `chromedriver` version 2.41 will be downloaded.
 * For beta versions, you'll have to require the beta version of `chromedriver`
 using `Webdrivers::Chromedriver.required_version`.

The gem looks for the Chrome/Chromium version that `chromedriver` will use by default.
You can override this behavior by providing a path to the browser binary you want to use:

```ruby
Selenium::WebDriver::Chrome.path = '/chromium/install/path/to/binary'
```

Alternatively, you can define the path via the `WD_CHROME_PATH` environment
variable.

This is also required if Google Chrome is not installed in its
[default location](https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver).

#### Chrome on Heroku

Follow the specific instructions [here](https://github.com/titusfortner/webdrivers/wiki/Heroku-buildpack-google-chrome) if you're using `heroku-buildpack-google-chrome`.

#### Microsoft Edge (Chromium)

Microsoft Edge (Chromium) support was added in v4.1.0. Notes
from the [Chrome/Chromium](https://github.com/titusfortner/webdrivers#chromechromium)
section apply to this browser as well.

Please note that `msedgedriver` requires `selenium-webdriver` v4.

#### WSLv1 support

While WSLv1 is not designed to run headful applications like Chrome, it can run exes; as such when found to be running
in WSL, `webdrivers` will use Chrome on the Windows filesystem.

It's recommended that you install the new PowerShell (PS7) to avoid [a known issue](https://github.com/microsoft/terminal/issues/367)
with the console font being changed when calling the old PowerShell (PS5).

#### WSLv2 support

Webdrivers will detect WSLv2 as running on Linux and use Chrome on the Linux filesystem.

WSLv2 doesn't support connecting to host ports out of the box, so it isn't possible to connect to Chromedriver on
Windows without extra configurations, see: https://github.com/microsoft/WSL/issues/4619. The simplest way to use 
Chromedriver with WSLv2 is to run Chrome headless on Linux.

#### Chrome and Edge on Apple M1 (`arm64`)

If you're switching from Intel to M1, you'll have to manually delete the existing Intel (`mac64`) driver before the 
M1 (`arm64`) build can be downloaded. Otherwise, you'll get an error: `Bad CPU type in executable - ~/.webdrivers/chromedriver (Errno::E086)`

## Wiki

Please see the [wiki](https://github.com/titusfortner/webdrivers/wiki)
for solutions to commonly reported issues.

Join us in the `#webdrivers-gem` channel on [Slack](https://seleniumhq.herokuapp.com/)
if you have any questions.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT),
see LICENSE.txt for full details and copyright.

## Contributing

Bug reports and pull requests are welcome [on GitHub](https://github.com/titusfortner/webdrivers).
Run `bundle exec rake` and squash the commits in your PRs.


CHANGELOG.md
# 5.3.1 (2023-07-31)
* Add post_install_message to explain status of gem

# 5.3.0 (2023-07-31)
* Add support for Chrome for Testing drivers ([#237](https://github.com/titusfortner/webdrivers/issues/249), thanks sadahiro-ono)
* Support is restricted to Selenium 4.0 - 4.10 to encourage people to update to Selenium 4.11

# 5.2.0 (2022-09-29)
* `chromedriver` - Fix downloading on M1 macs from version 106.0.5249.61 ([#237](https://github.com/titusfortner/webdrivers/issues/237), thanks jmccure)

# 5.1.0 (2022-09-19)
* Improve error messages (#232, thanks dan-jensen)
* Fix circular require bug with railtie (#233, thanks amatsuda)
* Support downloading correct geckodriver on Apple silicon (#234 thanks stephannv)
* Update linux locations for Chrome to latest chromedriver finder (thanks entretechno-jeremiah) 
* Add support for Microsoft Edge on Linux (#236)

# 5.0.0 (2021-10-19)
* Require Selenium 4+
* Remove custom Logger and use the one in Selenium 4

# 4.7.0 (2021-10-14)
* Add support for Selenium 4.x #218 (thanks yahonda )
* Allow Edge on Linux
* Update required Selenium > 3.141 because of a change in 4.6
* Require Ruby 2.6+
* Fix IE downloads with the change in location of assets

# 4.6.1 (2021-08-19)
* Fix bug in IEdriver caused by bad formatting in recent release

# 4.6.0 (2021-02-26)
* Update `Edgedriver` naming to support Selenium 4 beta (([#206](https://github.com/titusfortner/webdrivers/pull/206))

# 4.5.0 (2021-01-19)
* Improve format of README usage section. Thanks, [okuramasafumi](https://github.com/okuramasafumi)!
* `chromedriver` - Add support for Apple M1 specific builds ([#193](https://github.com/titusfortner/webdrivers/pull/193)). Thanks, [MichaelHoste](https://github.com/MichaelHoste)!
* `msedgedriver` - Add support for Apple M1 specific builds.
* `msedgedriver` - Added automatic selection of 32-bit vs 64-bit builds on Windows and WSLv1.  
* When using this gem in WSLv2, Chrome installed on the Linux filesystem will be automatically used ([#196](https://github.com/titusfortner/webdrivers/pull/196)). 
  WSLv1 continues to automatically use Chrome from the Windows filesystem though. Thanks, [fabioxgn](https://github.com/fabioxgn)!
  * We'll aim to make the behavior consistent in a future release.

### 4.4.2 (2020-12-29)
* Use new powershell if available to enhance WSL support ([#176](https://github.com/titusfortner/webdrivers/issues/176), 
  [#186](https://github.com/titusfortner/webdrivers/pull/186)). Thanks, [G-Rath](https://github.com/G-Rath) and 
  [131](https://github.com/131)!
* Update rubocop version (development dependency only) ([#178](https://github.com/titusfortner/webdrivers/pull/178)).
  Thanks, [utkarsh2102](https://github.com/utkarsh2102)!
* Replace `git ls-files` with `dir` in the gemspec for better Debian support ([#179](https://github.com/titusfortner/webdrivers/pull/179),
  [#184](https://github.com/titusfortner/webdrivers/issues/184)).
  Thanks again, [utkarsh2102](https://github.com/utkarsh2102)!
* Update README with `require: false` when adding gem to Gemfile ([#183](https://github.com/titusfortner/webdrivers/pull/183)).
* Make WSL detection more generic ([#187](https://github.com/titusfortner/webdrivers/pull/187)).
  [rbclark](https://github.com/rbclark)!
  
### 4.4.1 (2020-06-01)
* Do not include binstubs used for troubleshooting only. Fixes [#174](https://github.com/titusfortner/webdrivers/issues/174).

### 4.4.0 (2020-05-31)
* Implement support for `chromedriver` in Windows Subsystem for Linux (WSL) v1 ([#172](https://github.com/titusfortner/webdrivers/issues/172)). 
  Thanks, [G-Rath](https://github.com/G-Rath)!
* Chrome/Edgedriver - Fix [#171](https://github.com/titusfortner/webdrivers/issues/171) by making sure the cached
driver build version matches the browser build version before using it.
* Chrome/Edgedriver - Driver version check now matches the driver and browser `major.minor.build` versions instead of just 
the major version to be fully compliant with the `chromedriver` version selection
[document](https://chromedriver.chromium.org/downloads/version-selection).

### 4.3.0 (2020-04-14)
* Add support for Microsoft Edge (Chromium) stable
* Drop support for Ruby < 2.4.0

### 4.2.0 (2019-12-27)
* Add support for Microsoft Edge (Chromium) Beta releases ([#155](https://github.com/titusfortner/webdrivers/pull/155))
* Use tilde expansion to resolve user's home directory ([#166](https://github.com/titusfortner/webdrivers/pull/161))
* Add support for Chromium installed using Snap on Ubuntu ([#163](https://github.com/titusfortner/webdrivers/pull/163)). Thanks Tietew!

### 4.1.3 (2019-10-07)
* Require rubyzip version 1.3.0 or higher to fix [rubyzip#403](https://github.com/rubyzip/rubyzip/pull/403). Thanks rhymes! ([#153](https://github.com/titusfortner/webdrivers/pull/153))
* Fix a bug where the file deletion confirmation was printed even when there were no files to delete.

### 4.1.2 (2019-07-29)
* Fix a bug related to raising `BrowserNotFound`.

### 4.1.1 (2019-07-18)
* Raise `BrowserNotFound` if Chrome and Edge binary are not found (issue [#144](https://github.com/titusfortner/webdrivers/issues/144)).

### 4.1.0 (2019-07-03)
* Add support for `msedgedriver` (issue [#93](https://github.com/titusfortner/webdrivers/issues/93))
* Allow users to provide a custom binary path via `WD_CHROME_PATH` 
and `WD_EDGE_CHROME_ATH` environment variables (issues #[137](https://github.com/titusfortner/webdrivers/issues/137) 
and [#93](https://github.com/titusfortner/webdrivers/issues/93))
* Fix a bug where the user given Chrome binary path via `Selenium::WebDriver::Chrome.path` 
was not properly escaped (issue [#139](https://github.com/titusfortner/webdrivers/issues/139))
* Fix miscellaneous code warnings.
* ~~**Announcement**: As of 06/21/2019, `heroku-buildpack-google-chrome` 
no longer requires a [workaround](https://github.com/titusfortner/webdrivers/wiki/Heroku-buildpack-google-chrome) 

lib/webdrivers/common.rb
# frozen_string_literal: true

require 'rubygems/package'
require 'webdrivers/network'
require 'webdrivers/system'
require 'selenium-webdriver'
require 'webdrivers/logger'
require 'webdrivers/version'

module Webdrivers
  class ConnectionError < StandardError
  end

  class VersionError < StandardError
  end

  class NetworkError < StandardError
  end

  class BrowserNotFound < StandardError
  end

  DEFAULT_CACHE_TIME = 86_400 # 24 hours
  DEFAULT_INSTALL_DIR = File.expand_path('~/.webdrivers')

  class << self
    attr_accessor :proxy_addr, :proxy_port, :proxy_user, :proxy_pass
    attr_writer :install_dir, :cache_time

    #
    # Returns the amount of time (Seconds) the gem waits between two update checks.
    # @note Value from the environment variable "WD_CACHE_TIME" takes precedence over Webdrivers.cache_time. If neither
    # are set, it defaults to 86,400 Seconds (24 hours).
    #
    def cache_time
      @cache_time ||= (ENV['WD_CACHE_TIME'] || DEFAULT_CACHE_TIME)
      @cache_time.to_i
    end

    #
    # Returns the install (download) directory path for the drivers.
    #
    # @return [String]
    def install_dir
      @install_dir ||= ENV['WD_INSTALL_DIR'] || DEFAULT_INSTALL_DIR
    end

    def logger
      @logger ||= Webdrivers::Logger.new
    end

    #
    # Provides a convenient way to configure the gem.
    #
    # @example Configure proxy and cache_time
    #   Webdrivers.configure do |config|
    #     config.proxy_addr = 'myproxy_address.com'
    #     config.proxy_port = '8080'
    #     config.proxy_user = 'username'
    #     config.proxy_pass = 'password'
    #     config.cache_time = 604_800 # 7 days
    #   end
    #
    def configure
      yield self
    end

    def net_http_ssl_fix
      raise 'Webdrivers.net_http_ssl_fix is no longer available.' \
      ' Please see https://github.com/titusfortner/webdrivers#ssl_connect-errors.'
    end
  end

  class Common
    class << self
      attr_writer :required_version

      #
      # Returns the user defined required version.
      #
      # @return [Gem::Version]
      def required_version
        normalize_version(@required_version ||= nil)
      end

      #
      # Triggers an update check.
      #
      # @return [String] Path to the driver binary.
      def update
        return driver_path if user_defined_driver_path

        if correct_binary?
          msg = required_version != EMPTY_VERSION ?  'The required webdriver version' : 'A working webdriver version'
          Webdrivers.logger.debug "#{msg} is already on the system"
          return driver_path
        end

        remove
        System.download(download_url, driver_path)
      end

      #
      # Deletes the existing driver binary.
      #
      def remove
        @download_url = nil
        @latest_version = nil
        System.delete "#{System.install_dir}/#{file_name.gsub('.exe', '')}.version"
        System.delete driver_path
      end

      #
      # Returns path to the driver binary.
      #
      # @return [String]
      def driver_path
        user_defined_driver_path || File.absolute_path(File.join(System.install_dir, file_name))
      end

      private

      def user_defined_driver_path
        env_var_name = "WD_#{name.split('::').last.upcase}_PATH"

        return if ENV[env_var_name].nil?

        Webdrivers.logger.debug "#{env_var_name}: #{ENV[env_var_name]}"
        ENV[env_var_name]
      end

      def download_version
        required_version == EMPTY_VERSION ? latest_version : required_version
      end

      def download_url
        @download_url ||= direct_url(download_version).tap { |url| Webdrivers.logger.debug "#{file_name} URL: #{url}" }
      end

      def exists?
        System.exists? driver_path
      end

      def correct_binary?
        current_version == download_version
      rescue ConnectionError, VersionError
        driver_path if sufficient_binary?
      end

      def sufficient_binary?
        exists?
      end

      def normalize_version(version)
        Gem::Version.new(version.to_s)
      end

      def binary_version
        version = System.call(driver_path, '--version')
        Webdrivers.logger.debug "Current version of #{driver_path} is #{version}"
        version
      rescue Errno::ENOENT
        Webdrivers.logger.debug "No Such File or Directory: #{driver_path}"
        nil
      end

      # Returns cached driver version if cache is still valid and the driver binary exists.
      # Otherwise caches the given version (typically the latest available)
      # In case of Chrome, it also verifies that the driver build and browser build versions are compatible.
      # Example usage: lib/webdrivers/chromedriver.rb:34
      def with_cache(file_name, driver_build = nil, browser_build = nil)
        if System.valid_cache?(file_name) && exists? && (driver_build == browser_build)
          cached_version = System.cached_version(file_name)
          Webdrivers.logger.debug "using cached version as latest: #{cached_version}"
          normalize_version cached_version
        else
          version = yield
          System.cache_version(file_name, version)
          normalize_version version
        end
      end

      EMPTY_VERSION = Gem::Version.new('')
    end
  end
end
